### PR TITLE
fix(cli): fast resolver support all mainFields in package.json

### LINKED ISSUE #29229: EXPO_USE_FAST_RESOLVER=1 ignores resolverMainFields config
### Summary

We tried enable `EXPO_USE_FAST_RESOLVER=1` but it seems it ignores/prioritize `browser` over all other fields even over `main` field. Check following example of `package.json`:

```json
    "main": "src/index.ts",
    "browser": {
        "./src/index": "./src/index-browser.ts",
        "./src/index.ts": "./src/index-browser.ts",
    },
    "react-native": {
        "./src/index": "./src/index.ts",
        "./src/index.ts": "./src/index.ts",
    },
```

With this configuration metro resolver will prioritize `browser` over everything else. I tried to override `resolverMainFields` in `metro.config.js` but it seems to be ignored too. Only thing that works is removing entrire `browser` part from that `package.json`

### What platform(s) does this occur on?

Android, iOS

### SDK Version

51

### Environment

  expo-env-info 1.2.0 environment info:
    System:
      OS: macOS 13.3.1
      Shell: 5.9 - /bin/zsh
    Binaries:
      Node: 18.17.1 - ~/Library/Caches/fnm_multishells/8610_1716809312259/bin/node
      Yarn: 4.1.1 - ~/Library/Caches/fnm_multishells/8610_1716809312259/bin/yarn
      npm: 9.6.7 - ~/Library/Caches/fnm_multishells/8610_1716809312259/bin/npm
      Watchman: 2024.05.06.00 - /usr/local/bin/watchman
    Managers:
      CocoaPods: 1.15.2 - /usr/local/bin/pod
    SDKs:
      iOS SDK:
        Platforms: DriverKit 22.4, iOS 16.4, macOS 13.3, tvOS 16.4, watchOS 9.4
      Android SDK:
        Android NDK: 26.2.11394342
    IDEs:
      Android Studio: 2022.3 AI-223.8836.35.2231.10811636
      Xcode: 14.3.1/14E300c - /usr/bin/xcodebuild
    npmGlobalPackages:
      eas-cli: 7.8.5
      expo-cli: 6.3.10
    Expo Workflow: bare

### Minimal reproducible example

1. Install some package that uses `browser` field in `package.json` (for us it is `@trezor/connect`)
2. Run `EXPO_USE_FAST_RESOLVER=1 npx expo`
3. It will resolve `browser` version of package instead of `react-native`

### PR DESCRIPTION
**Pull request by Nodonisko**
# Why

If there was defined overrides for inside package paths in package.json `react-native` field it was ignored by resolver. This fix should use `mainFields` config instead of hardcoded `browser` to resolve that path.

For example following `package.json` always used `browser` mapping instead of `react-native`

```json
"main": "src/index.ts",
    "browser": {
        "./src/index": "./src/index-browser.ts",
        "./src/index.ts": "./src/index-browser.ts",
    },
    "react-native": {
        "./src/index": "./src/index.ts",
        "./src/index.ts": "./src/index.ts",
    },
```

Resolves #29229 

# How

# Test Plan

- Added unit tests
- Tested manually

# Checklist

- [x] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [x] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [x] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

**Comment by Nodonisko** [comment]
Hi @EvanBacon can you please take a look? This is kind of blocker for us to use fast resolver. Thanks.

File: packages/@expo/cli/src/start/server/metro/createExpoMetroResolver.ts [hunk 1]
```diff
@@ -180,15 +182,24 @@ export function createFastResolver({
...
             ? undefined
-            : // Enable `browser` field support
+            : // Enable package.json `browser` and `react-native` fields support
               (pkg: any, _resolvedPath: string, relativePathIn: string): string => {
                 let relativePath = relativePathIn;
                 if (relativePath[0] !== '.') {
                   relativePath = `./${relativePath}`;
                 }
 
-                const replacements = pkg.browser;
```
Comment (EvanBacon) [inline packages/@expo/cli/src/start/server/metro/createExpoMetroResolver.ts:190]: `browser` field and `react-native` fields aren't similar afaict. [spec](nctzombie/package-browser-field-spec). The `mainFields` don't appear to have the same capabilities as the browser field.

Comment (Nodonisko) [reply packages/@expo/cli/src/start/server/metro/createExpoMetroResolver.ts:190]: @EvanBacon Link doesn't seem to work (404)

I tried to copy behaviour of default Metro resolver which I believe shouldn't be modified it this should be drop-in replacement.

**Comment by Nodonisko** [comment]
This seems to be fixed in latest Metro version that should include this resolver.

**Pull request closed without merging**